### DISCUSSION TITLE
bad completions

### DISCUSSION THREAD

**Discussion by nbecker** [minad/vertico#462] [Q&A]
I'm new to vertico (been using emacs for 40 years).

One problem I have has got me ready to abandon vertico, but maybe you can help me live with vertico.

I have a directory call s-band, and another called s-band-old.  Everytime I do C-x C-f s-band, it autocompletes to s-band-old and opens that when I (muscle memory) hit RET.  This is driving me nuts.  Is there a setting that would fix this?  I see no logic to choosing 's-band-old' when I typed 's-band' when a valid completion called 's-band' exists.

**Comment by minad** [comment]
Hello! I don't see this problem. Can you reproduce it in emacs -Q? Do you use some special completion style settings?

**Comment by nbecker** [comment]
On Wed, Mar 27, 2024 at 10:11 AM Daniel Mendler ***@***.***>
wrote:

> Hello! I don't see this problem. Can you reproduce it in emacs -Q? Do you
> use some special completion style settings?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/minad/vertico/discussions/462#discussioncomment-8929036>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4VUDFWCY3DGF3KF62EEDY2LARBAVCNFSM6AAAAABFK4BG6WVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DSMRZGAZTM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
Thanks for the quick response!  The only vertico setting I have are:
;; Enable vertico
(use-package vertico
  :init
  (vertico-mode)
  :custom
  (vertico-sort-function 'vertico-sort-history-alpha))
;; Enable richer annotations using the Marginalia package
  (use-package marginalia
    :bind (:map minibuffer-local-map
  ("M-A". marginalia-cycle))
    :init
    (marginalia-mode))

**Reply by minad** [reply]
When I enter C-x C-f s-band it will put s-band at the top since that's an exactly matching candidate. When you type C-x C-f s-b TAB (or RET) then it will complete to the topmost candidate given that partial input. This could be s-band-old if that's the last visited file in the file history.

**Comment by nbecker** [comment]
On Wed, Mar 27, 2024 at 10:32 AM Daniel Mendler ***@***.***>
wrote:

> When I enter C-x C-f s-band it will put s-band at the top since that's an
> exactly matching candidate. When you type C-x C-f s-b TAB (or RET) then it
> will complete to the topmost candidate given that partial input. This could
> be s-band-old if that's the last visited file in the file history.
>
I'm having trouble intentionally reproducing this now.  I have to figure
out how I'm triggering it.  I has happened so many times I can't count, but
not now when I'm trying to debug it :)

**Comment by nbecker** [comment]
OK, I think I got it.  If I type C-x C-f ~/s-ba[TAB] I will get ~/s-band.
But if I'm currently in the directory ~/Download and then type ../~/s-band
look what happens, I got the wrong completion:

On Wed, Mar 27, 2024 at 10:47 AM Neal Becker ***@***.***> wrote:

>
>
> On Wed, Mar 27, 2024 at 10:32 AM Daniel Mendler ***@***.***>
> wrote:
>
>> When I enter C-x C-f s-band it will put s-band at the top since that's an
>> exactly matching candidate. When you type C-x C-f s-b TAB (or RET) then it
>> will complete to the topmost candidate given that partial input. This could
>> be s-band-old if that's the last visited file in the file history.
>>
> I'm having trouble intentionally reproducing this now.  I have to figure
> out how I'm triggering it.  I has happened so many times I can't count, but
> not now when I'm trying to debug it :)
>


-- 
*Those who don't understand recursion are doomed to repeat it*

**Reply by minad** [reply]
I haven't checked but this could be an Emacs bug regarding the completion of shadowed paths (this is a buggy area). I use the following configuration, to automatically clean up shadowed paths (see also the readme):

~~~ elisp
:hook (rfn-eshadow-update-overlay . vertico-directory-tidy)
~~~

**Comment by nbecker** [comment]
I modified my init to:
(use-package vertico
  :init
  (vertico-mode)
  :hook (rfn-eshadow-update-overlay . vertico-directory-tidy)
  :custom
  (vertico-sort-function 'vertico-sort-history-alpha))
;; Enable richer annotations using the Marginalia package
  (use-package marginalia
    :bind (:map minibuffer-local-map
  ("M-A". marginalia-cycle))
    :init
    (marginalia-mode))
Hopefully that's right, I'm not that familiar with use-package.

Doesn't fix the problem completely.  If I visit ~/Downloads, the if I enter
C-x C-f (here ~/Downloads is offered)../s-band it still completes to
~/Downloads/../s-band-old.  If I instead just type ~/s-band, that works.

On Wed, Mar 27, 2024 at 5:41 PM Daniel Mendler ***@***.***>
wrote:

> I haven't checked but this could be an Emacs bug regarding the completion
> of shadowed paths (this is a buggy area). I use the following
> configuration, to automatically clean up shadowed paths (see also the
> readme):
>
> :hook (rfn-eshadow-update-overlay . vertico-directory-tidy)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/minad/vertico/discussions/462#discussioncomment-8933718>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4VUCIZP46NUI4PX455Q3Y2MVHXAVCNFSM6AAAAABFK4BG6WVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DSMZTG4YTQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
*Those who don't understand recursion are doomed to repeat it*

**Reply by minad** [reply]
> Doesn't fix the problem completely.  If I visit ~/Downloads, the if I enter
C-x C-f (here ~/Downloads is offered)../s-band it still completes to
~/Downloads/../s-band-old.  If I instead just type ~/s-band, that works.

I don't see this. I see the correct behavior as described in https://github.com/minad/vertico/discussions/462#discussioncomment-8929321.

**Reply by minad** [reply]
Maybe there is some misunderstanding? Could you please tell me your exact key presses? What do you expect to happen precisely? Vertico always completes the selected candidate (by default the top most). If this is undesired, press up/down. Then there is also the possibility to expand the input to the common prefix of the candidates, but this is default completion behavior and not the behavior of Vertico. See also https://github.com/minad/vertico?tab=readme-ov-file#completion-styles-and-tab-completion.
